File: ios/HarnessMobile/Chat/ChatTemporalContext.swift
import Foundation

enum ChatTemporalContext {
    private static let sentAtPrefix = #"^\[sent_at=[^\]]+\]\n"#

    static func temporalContextBlock(now: Date = Date(), timeZone: TimeZone = .current) -> String {
        let tz = timeZone.identifier
        let formatted = formatCurrentDateTime(now, timeZone: timeZone)
        return """
        [TEMPORAL_CONTEXT]
        Current local date and time (\(tz)): \(formatted)
        When present, a message begins with [sent_at=...] (ISO 8601 UTC) for when it was sent.
        Use sent_at together with the current time above to interpret relative dates and whether discussed future plans, events, or deadlines have already passed.
        """
    }

    static func annotateMessageContentForModel(_ content: String, timestampMs: Int64?) -> String {
        guard let timestampMs else { return content }
        if content.range(of: sentAtPrefix, options: .regularExpression) != nil {
            return content
        }
        let sentAt = ISO8601DateFormatter().string(from: Date(timeIntervalSince1970: Double(timestampMs) / 1000))
        return "[sent_at=\(sentAt)]\n\(content)"
    }

    private static func formatCurrentDateTime(_ date: Date, timeZone: TimeZone) -> String {
        let formatter = DateFormatter()
        formatter.timeZone = timeZone
        formatter.dateStyle = .full
        formatter.timeStyle = .long
        return formatter.string(from: date)
    }
}

File: src/shared/chatTemporalContext.ts
/** Prefix on model-facing message bodies; not stored on disk. */
const SENT_AT_LINE = /^\[sent_at=[^\]]+\]\n/;

function resolveTimeZone(timeZone?: string): string {
  const tz = (timeZone ?? "").trim();
  if (tz) {
    try {
      new Intl.DateTimeFormat("en-US", { timeZone: tz }).format(new Date());
      return tz;
    } catch {
      // fall through
    }
  }
  return Intl.DateTimeFormat().resolvedOptions().timeZone;
}

/** Human-readable "now" for the system prompt (local timezone). */
export function formatCurrentDateTimeForPrompt(now: Date = new Date(), timeZone?: string): string {
  const tz = resolveTimeZone(timeZone);
  return new Intl.DateTimeFormat(undefined, {
    timeZone: tz,
    dateStyle: "full",
    timeStyle: "long",
  }).format(now);
}

export function formatTemporalContextBlock(now: Date = new Date(), timeZone?: string): string {
  const tz = resolveTimeZone(timeZone);
  const formatted = formatCurrentDateTimeForPrompt(now, tz);
  return [
    "[TEMPORAL_CONTEXT]",
    `Current local date and time (${tz}): ${formatted}`,
    "When present, a message begins with [sent_at=...] (ISO 8601 UTC) for when it was sent.",
    "Use sent_at together with the current time above to interpret relative dates and whether discussed future plans, events, or deadlines have already passed.",
  ].join("\n");
}

/** Annotate message text for the model only; leaves stored content unchanged. */
export function annotateMessageContentForModel(content: string, timestampMs?: number): string {
  if (timestampMs == null || !Number.isFinite(timestampMs)) return content;
  if (SENT_AT_LINE.test(content)) return content;
  const sentAt = new Date(timestampMs).toISOString();
  return `[sent_at=${sentAt}]\n${content}`;
}

export function stripSentAtPrefix(content: string): string {
  return content.replace(SENT_AT_LINE, "");
}
